plugin.py

```diff
@@ -1,62 +1,69 @@
 import os
 import hashlib
 import shutil
-import threading
 import urllib.request
 import sublime
 
-from sublime_lib import ActivityIndicator
+from LSP.plugin import AbstractPlugin
+from LSP.plugin import ClientConfig
+from LSP.plugin import register_plugin  # https://github.com/sublimelsp/LSP/issues/899
+from LSP.plugin import unregister_plugin  # https://github.com/sublimelsp/LSP/issues/899
+from LSP.plugin import WorkspaceFolder
+from LSP.plugin.core.typing import List, Optional, Dict
 
-from LSP.plugin.core.handlers import LanguageHandler
-from LSP.plugin.core.settings import ClientConfig, read_client_config
 
+class Lemminx(AbstractPlugin):
 
-def is_java_installed() -> bool:
-    return shutil.which("java") is not None
+    binary = None  # type: Optional[str]
+    checksum = None  # type: Optional[str]
+    url = None  # type: Optional[str]
+    version = None  # type: Optional[str]
 
+    @classmethod
+    def name(cls) -> str:
+        return 'lemminx'
 
-def package_cache() -> str:
-    cache_path = os.path.join(sublime.cache_path(), __package__)
-    os.makedirs(cache_path, exist_ok=True)
-    return cache_path
-
-
-def merge_configs(target: dict, source: dict):
-    for key, value in target.items():
-        new_value = source.get(key)
-        if new_value is not None:
-            if value:
-                if isinstance(value, dict):
-                    yield key, dict(merge_configs(value, new_value))
-                    continue
-
-                if isinstance(value, list):
-                    if isinstance(value[0], dict):
-                        # dicts are unhashable
-                        yield key, new_value
-                        continue
-                    # always merge user values into defaults
-                    yield key, list(set(value) + set(new_value))
-                    continue
-
-            yield key, new_value
-
-        else:
-            yield key, value
+    @classmethod
+    def additional_variables(cls) -> Optional[Dict[str, str]]:
+        """
+        Additional template variables for expansion in
+        LSP-lemminx.sublime-settings
+        """
+        assert isinstance(cls.version, str)
+        return {'version': cls.version}
 
-    return None
+    @classmethod
+    def needs_update_or_installation(cls) -> bool:
+        # mega hack, remove! https://github.com/sublimelsp/LSP/issues/899
+        plugin_loaded()
```

Comment:
It feels odd to run the `plugin_loaded()` from within the class which then fills attributes of that class.

Those initialization and cleanup methods should keep static class methods.

Reply:
You're completely right, but I would like to do that in another pull request.

Reply:
And by another pull request I mean I don't want to fix it in sublimelsp/LSP#979 as it's getting pretty large. I made an attempt at fixing it but the amount of global state that's not properly cleaned up is so vast that it needs another pull request.